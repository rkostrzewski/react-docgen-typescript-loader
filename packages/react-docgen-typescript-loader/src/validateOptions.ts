import validate from "@webpack-contrib/schema-utils";
import LoaderOptions from "./LoaderOptions";

const schema = {
  type: "object",
  additionalProperties: false,
  properties: {
    skipPropsWithName: {
      anyOf: [
        { type: "string", minLength: 1 },
        { type: "array", minItems: 1, items: { type: "string", minLength: 1 } },
      ],
    },

    skipPropsWithoutDoc: {
      type: "boolean",
    },

    tsconfigPath: {
      type: "string",
      minLength: 1,
    },

    compilerOptions: {
      type: "object",
    },

    docgenCollectionName: {
      anyOf: [{ type: "string", minLength: 1 }, { type: "null" }],
    },

    setDisplayName: {
      type: "boolean",
    },

    includes: {
      type: "array",
      items: {
        type: "string",
      },
    },

    excludes: {
      type: "array",
      items: {
        type: "string",
      },
    },
  },
};

function validateOptions(options: LoaderOptions = {}) {
  validate({
    name: "react-docgen-typescript-loader",
    schema,
    target: options,
  });
}

export default validateOptions;
