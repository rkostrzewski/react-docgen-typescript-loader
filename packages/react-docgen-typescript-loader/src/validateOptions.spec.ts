import validateOptions from "./validateOptions";

it("throws error on unexpected field", () => {
  expect(() =>
    validateOptions({ invalidField: "fail" } as any),
  ).toThrowErrorMatchingSnapshot();
});

it("does not throw on empty", () => {
  expect(() => validateOptions()).not.toThrow();
});

describe("skipPropsWithName", () => {
  it("rejects empty string", () => {
    expect(() =>
      validateOptions({ skipPropsWithName: "" }),
    ).toThrowErrorMatchingSnapshot();

    expect(() => validateOptions({ skipPropsWithName: "prop" })).not.toThrow();
  });

  it("rejects empty array", () => {
    expect(() =>
      validateOptions({ skipPropsWithName: [] }),
    ).toThrowErrorMatchingSnapshot();

    expect(() =>
      validateOptions({ skipPropsWithName: ["prop"] }),
    ).not.toThrow();
  });
});

describe("compilerOptions", () => {
  it("accepts object of any shape", () => {
    expect(() =>
      validateOptions({
        compilerOptions: {
          option: "test",
          otherOption: { secondField: "test" },
        },
      } as any),
    ).not.toThrow();
  });
});

describe("v2 includes/excludes fields", () => {
  it("throws error if included", () => {
    expect(() =>
      validateOptions({
        includes: ["*\\.stories\\.tsx$"],
      } as any),
    ).toThrowErrorMatchingSnapshot();
  });
});
