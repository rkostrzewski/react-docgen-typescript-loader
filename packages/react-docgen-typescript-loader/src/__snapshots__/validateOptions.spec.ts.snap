// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`skipPropsWithName rejects empty array 1`] = `
"react-docgen-typescript-loader

  Options Validation Error

  options.skipPropsWithName  should match some schema in anyOf
  options.skipPropsWithName  should NOT have less than 1 items
  options.skipPropsWithName  should be string
"
`;

exports[`skipPropsWithName rejects empty string 1`] = `
"react-docgen-typescript-loader

  Options Validation Error

  options.skipPropsWithName  should match some schema in anyOf
  options.skipPropsWithName  should be array
  options.skipPropsWithName  should NOT be shorter than 1 characters
"
`;

exports[`throws error on unexpected field 1`] = `
"react-docgen-typescript-loader

  Options Validation Error

  options['invalidField']  is an invalid additional property
"
`;

exports[`v2 includes/excludes fields throws error if included 1`] = `
"react-docgen-typescript-loader

  Options Validation Error

  options['includes']  is an invalid additional property
"
`;
